js/turtleactions: refactor, lint & prettify

ricknjacky opened the pull request:
issue references: #2609, #2767

ricknjacky commented:
@meganindya pls review this.

ksraj123 commented on js/turtleactions/IntervalsActions.js:
```diff
@@ -19,6 +19,12 @@
+/*exported setupIntervalsActions*/
+
+/* global MUSICALMODES, _, getNote, Singer, getModeLength, logo, NOINPUTERRORMSG, blocks, 
+MusicBlocks, Mouse, args
```
are you sure `args` is a global here? where does it come from? Usually args contains the arguments for a block passed into the flow function. Asking as it seems to be a reason for a few bugs in some of the functions in this file.

ricknjacky replied:
yes it is global:- please use this [link ](https://github.com/sugarlabs/musicblocks/blob/d081fe3ef8e4ef14c503a271dd7b417f8802ac7c/js/js-export/API/MeterBlocksAPI.js#L35)for your perusal

ksraj123 replied:
@ricknjacky For the link you have shared, I think `args` is local to the `setBPM` function. Almost all functions in the `MeterBlockAPI` class are declaring local variable `args` in them. Also, I think all globals could be accessed in the browser console which args cannot. Also, in the `intervalActions.js` file there is only one instance of `args` in the `defineMode` function which I think is an error as when you try to use the blocks that make use of that function you get a console error about it. I have also opened a PR regarding it. cc - @meganindya

meganindya replied:
Where is `args`?

ksraj123 commented on js/turtleactions/MeterActions.js:
```diff
@@ -16,6 +16,12 @@
 
+/*exported setupMeterActions*/
+
+/* global rationalToFraction, _, blk, Singer, _blk, logo, TONEBPM, blocks, 
```
Are you sure `blk` and specially `_blk` are globals? Where do they come from? Not everything which is not defined is necessarily a global. Thanks

ricknjacky replied:
Yes it is global, use this [link ](https://github.com/sugarlabs/musicblocks/blob/d081fe3ef8e4ef14c503a271dd7b417f8802ac7c/js/activity.js#L3413)for your perusal. 

and _blk is I presume for translation purposes. Using _ is used throughout codebase for the same purposes.

and yes, I concur with your statement 
>Not everything which is not defined is necessarily a global.

hope I have answered your queries.

ksraj123 replied:
By _ I think you mean the _ function in the `js/utils/utils.js` file. It is a function which takes in strings as arguments, I can't see how you are relating _blk to it. Also for the link you shared we can clearly see that the `errorMsg` function takes in `blk` as an argument and so does almost every other function in the `MeterActions.js` file. `blk` actually is the index of the block in `blocks.blockList` so there is no way it can be global. cc - @meganindya

meganindya replied:
`blk` must have been passed to the methods from corresponding classes in `/blocks`. If it isn't that's an error. Certainly not a global.

ksraj123 commented on js/turtleactions/ToneActions.js:
```diff
@@ -16,6 +16,10 @@
 
+/*exported setupToneActions*/
+
+/* global Singer, logo, blocks, MusicBlocks, Mouse, last, _, VOICENAMES, instrumentsEffects, NOINPUTERRORMSG, arg */
```
Same here for `arg`

ricknjacky replied:
Yes, it is a global, I can be wrong but in this case arg is a global present in js/blocks sub-directory in multiple files in it. 

In this file, i.e. `js/turtleactions/ToneActions.js`  it is used inside the method for:- Frequency Modulator used to define a timbre.

meganindya replied:
Not a global. This is an error I mistakenly introduced while porting to `turtleactions`. See f24a94bda391539f997308ad87b30d09c30f0f86. Fix it pls.

ksraj123 reviewed (commented):
It would be more helpful if we had where the globals come from in the same file. I think the reason behind enabling linting was that we could identify and fix other issues in the process. I guess, fixing linting issues in a way that just make the errors go away is as good as not having linting at all. Just my humble opinion. Thanks

ricknjacky commented:
> It would be more helpful if we had where the globals come from in the same file. I think the reason behind enabling linting was that we could identify and fix other issues in the process. I guess, fixing linting issues in a way that just make the errors go away is as good as not having linting at all. Just my humble opinion. Thanks

Totally agree. But that's not what I did. 
With the nascent experience I have of working on the codebase, I have addressed your concerns.
Also, 

Core of MusicBlocks resides in the `/js`, `/blocks` and the `/turtleactions` subdirectories.  As a result there are many globals in these directories, `js/turtleactions` in this case. There are architectural issues pertaining to this as well, hopefully in MB2 we can obviate this instead of having this galore of GOD objects engendering regressions and all sorts of other issues.

Thanks a lot for the review btw

ksraj123 commented:
@meganindya Please go through my comments in my review which have been marked as resolved. I was working on some other issue that involved one of the files in this PR so I thought reviewing this would be a good opportunity to learn. Thanks

ricknjacky commented:
@ksraj123 Thx for pointing out your concerns. I have unresolved your comments on the review, waiting for a better answers to these. Surely it's a great opportunity to learn more about how the codebase is inter-connected.

meganindya commented on js/turtleactions/MeterActions.js:
```diff
@@ -373,6 +379,7 @@ function setupMeterActions() {
         }
 
         static getBeatFactor(turtle) {
+            // eslint-disable-next-line max-len
```
pls don't disable rules unless extraordinary circumstances. This one exceeds 100 characters, hence the warning.

meganindya reviewed (changes requested):
Add the locations too
```javascript
/*
   Global locations
    path..
        .., .., ...
 */
```

ricknjacky commented:
Thanks for pointing out my mistakes and sharing your feedback @meganindya 
I had couple of queries.

>blk must have been passed to the methods from corresponding classes in /blocks. If it isn't that's an error. Certainly not a global

>Not a global. This is an error I mistakenly introduced while porting to turtleactions. See f24a94b. Fix it pls.


How do I troubleshoot these? remove the variable itself or is there some other way round.

ricknjacky commented:
>Where is args?    

```
static defineMode(name, turtle, blk) {
            const tur = logo.turtles.ithTurtle(turtle);

            tur.singer.inDefineMode = true;
            tur.singer.defineMode = [];
            let modeName;
            if (name === null) {
                logo.errorMsg(NOINPUTERRORMSG, blk);
                modeName = "custom";
            } else {
                modeName = args[0].toLowerCase();
            }
```

ricknjacky commented:
>blk must have been passed to the methods from corresponding classes in /blocks. If it isn't that's an error. Certainly not a global.

@meganindya 
I have removed the argument `blk` and `_blk` accordingly.

Pls review and provide feedback, thanks.
